DOCSP-41957 - Mongo Client (#106)
mongodb/docs-php-library: source/connect/client.txt

.. _php-client:

=======================
Create a MongoDB Client
=======================

.. facet::
   :name: genre
   :values: reference

.. meta::
   :keywords: connection string, URI, server, Atlas, settings

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

Overview
--------

> To connect to a MongoDB deployment, you need two things:

Reviewer: ```suggestion
To connect to a MongoDB deployment, you must create the following items:
```

> - A **connection URI**, also known as a *connection string*, which tells the {+driver-short+}
>   which MongoDB deployment to connect to.
> - A **MongoDB\\Client** object, which creates the connection to the MongoDB deployment
>   and lets you perform operations on it.

Reviewer: ```suggestion
- **Connection URI**, also known as a *connection string*, which tells the {+driver-short+}
  which MongoDB deployment to connect to.
- **MongoDB\\Client** object, which creates the connection to the MongoDB deployment
  and lets you perform operations on it.
```

>   and lets you perform operations on it.

Reviewer: Q: are there two \\ because it's being escaped? It looks normal on the frontend.

Author: yeah, you need to do this when it's bold for some reason

> You can also use either of these components to customize the way the {+driver-short+} behaves
> while connected to MongoDB.

Reviewer: S: maybe you can add some more information about what this suggests. Maybe:

```suggestion
You can also set options within either or both of these components to
customize the way that the {+driver-short+} behaves
while connected to MongoDB.
```

This guide shows you how to create a connection string and use a ``MongoDB\Client`` object

> to connect to MongoDB.

Reviewer: Q: does this guide show how to create a connection string or does it just teach about connection strings?

Author: it describes the components of the connection string

Author: Do you think we need to be clearer about this point? I hoped that if we showed users all of the pieces of the connection string, they could put one together...?

Reviewer: I dont think many users are creating a connection string from its constituent pieces. Perhaps this sentence could express something more like:
```suggestion
This guide describes the components of a connection string and shows how to
 use a ``MongoDB\Client`` object to connect to a MongoDB deployment.
```

.. _php-connection-uri:

Connection URI
--------------

A standard connection string includes the following components:

.. list-table::
   :widths: 20 80
   :header-rows: 1

   * - Component
     - Description

   * - ``mongodb://``

     - Required. A prefix that identifies this as a string in the
       standard connection format.

>    * - ``username:password``

Reviewer: S: change to ``db_username`` and ``db_password``

Author: do these need to be disambiguated from other credentials? the following paragraph also says they're for the DB.

Reviewer: I think it could be either. Just figured we might start integrating those placeholders where applicable so that there are fewer mis-matches

Reviewer: Have you considered including that connection string image on this page and using those placeholders instead? another option

Author: Not opposed to this change, but should we be consistent with the Server manual?

https://www.mongodb.com/docs/manual/reference/connection-string/#connection-string-components

Reviewer 2: I'm a bit late here, but I don't think these should be `db_username` and `db_password`. DB-backed auth is but one mechanism. These could be credentials for a different auth mechanism, so more generic names are appropriate.

See also: [connection string spec](https://github.com/mongodb/specifications/blob/master/source/connection-string/connection-string-spec.md)

     - Optional. Authentication credentials. If you include these, the client
       authenticates the user against the database specified in ``authSource``.
       For more information about the ``authSource`` connection option, see

>        :ref:`php-auth`. 

Author: broken for now

   * - ``host[:port]``

     - Required. The host and optional port number where MongoDB is running. If you don't
       include the port number, the driver uses the default port, ``27017``.

   * - ``/defaultauthdb``
     
     - Optional. The authentication database to use if the
       connection string includes ``username:password@``
       authentication credentials but not the ``authSource`` option. If you don't include 
       this component, the client authenticates the user against the ``admin`` database.

   * - ``?<options>``

     - Optional. A query string that specifies connection-specific
       options as ``<name>=<value>`` pairs. See

>        :ref:`php-connection-options` for a full description of

Author: broken for now

       these options.

To learn more about connection strings, see
:manual:`Connection Strings </reference/connection-string>` in the

> MongoDB Server documentation.

Reviewer: ```suggestion
Server manual.
```

> For more information about creating a connection string, see
> :manual:`Connection Strings </reference/connection-string>` in the
> MongoDB Server documentation.

Reviewer: ```suggestion
To learn more about connection strings, see
:manual:`Connection Strings </reference/connection-string>` in the
MongoDB Server documentation.
```

> Create a ``MongoDB\Client``

Reviewer: S: avoid using monospace in titles: https://www.mongodb.com/docs/meta/style-guide/style/titles-and-headings/#:~:text=Don%27t%20apply%20font%20treatments%20(bold%2C%20italics%2C%20or%20monospace)%20to%20text%20in%20a%20title%20or%20heading.

Author: I'll follow the Style Guide, but this is a Bad Rule imo

---------------------------

> To create a connection to MongoDB, pass your connection string when constructing
>  an instance of the ``MongoDB\Client`` class.

Reviewer: I: something about the indentation is causing this to render weirdly:
```suggestion
To create a connection to MongoDB, pass your connection string when constructing
an instance of the ``MongoDB\Client`` class.
```

> To create a connection to MongoDB, construct an instance of the ``MongoDB\Client`` class,
> passing the connection URI as a string to the constructor.

Reviewer: ```suggestion
To create a connection to MongoDB, pass your connection string when constructing
 an instance of the ``MongoDB\Client`` class.
```

In the following example, the library uses a sample connection URI to connect to a MongoDB 
deployment on port ``27017`` of ``localhost``:

.. literalinclude:: /includes/connect/client.php
   :language: php
   :copyable: true

.. tip:: Reuse Your Client

   Because each ``MongoDB\Client`` object represents a pool of connections to the
   database, most applications require only a single instance of

>    ``MongoDB\Client``, even across multiple requests.

Reviewer: ```suggestion
   ``MongoDB\Client`` that you can use across multiple requests.
```

>    database, you can use a single instance of the class across multiple requests.

Reviewer 3: To be checked by @jmikola @alcaeus, but I don't think there's a pool of connections, since requests are blocking and PHP is monothread.

Reviewer 4: That is correct, PHP does not implement a connection pool. We have an [article in the extension docs](https://www.php.net/manual/en/mongodb.connection-handling.php) that details the connection persistence mechanism used by the PHP driver (current and the legacy one used before 1.2.0).

API Documentation
-----------------

To learn more about creating a ``MongoDB\Client`` object in the {+driver-short+},
see the following API documentation:

> - :ref:`MongoDB\Client <php-api-mongodbclient>`

Author: broken for now, possibly because this is staging

> - :ref:`MongoDB\Client <php-api-mongodbclient>`

Reviewer: I: this is presented as an API link, but I see a ref, is this supposed to link here https://www.mongodb.com/docs/php-library/current/reference/class/MongoDBClient/?

Author: yes, the PHP docs host their own API docs

Reviewer: Hmm, inter-repo refs should work. Maybe you can double check the snooty-toml to make sure its included in the intersphinx?

Author: Good thought, but this is within the same repo. Looks like my addition of the ref at the top of that file wasn't recognized, even though I had committed the change (I removed the last line of the file just so VS Code would let me save/commit again).

Reviewer: Looks to be working now!
